src/avx2/ycbcr.rs

```diff
@@ -72,7 +81,10 @@ macro_rules! ycbcr_image_avx2 {
                     let y = _mm256_add_epi32(_mm256_add_epi32(yr, yg), yb);
                     let y = _mm256_add_epi32(y, _mm256_set1_epi32(0x7FFF));
                     let y = _mm256_srli_epi32(y, 16);
-                    let y: [i32; 8] = core::mem::transmute(y);
+                    let y: [i32; 8] = avx_as_i32_array(y);
+                    let mut y: [u8; 8] = y.map(|x| x as u8);
+                    y.reverse();
+                    y_buffer.extend_from_slice(&y);
```

Comment:
```suggestion
                    y_buffer.extend(avx_as_i32_array(y).into_iter().map(|x| x as u8));
```
Using `_mm256_set_epi32` in `fill_buffers_avx2` and all these reverses were a little suspicious to me.

If you use `_mm256_setr_epi32` instead, you should be able to remove all these reverse calls.
I did that and it seems to pass all the tests 
[avx_fix.patch](https://github.com/user-attachments/files/23249275/avx_fix.patch)

Reply:
I've tried it and the change is functionally correct but regresses benchmarks by about 2% on my Zen4 machine. I do not know why.

Reply:
Sorry, I couldn't run the benchmarks before to test it. I also see a similar 2% regression too.
I checked the assembly that my suggestion produced and it wasn't great, more on that at the bottom.

That did get me to dig deeper and realize we can get rid of the unsafe `transmute` cast, ~~along with some improvements of 5-15% on the benches.
[avx_remove_transmute.patch](https://github.com/user-attachments/files/23297364/avx_remove_transmute.patch)~~
_(edit: this patch still needed another i32x8 permute to be correct)_

You can see in this snippet if you search `bswap` on the left, the i32 to u8 conversion is scalarized.
This PR on the left, ~~my incorrect patch with new SIMD on the right~~
https://rust.godbolt.org/z/39sxM6adh


<details><summary>Analyzing my earlier suggestion, it looks like LLVM struggles with the sequence of shuffling and cross-lane permutation that we want.</summary>


I believe my suggestion was a regression because it extracted the bytes only to reinsert them again to write 8 bytes at once. I'll probably have to write this in C and report it upstream to LLVM. This implicit truncating cast from i32 to u8 should be handled better unless it's a Rust issue.
```asm
vextracti128    xmm1, ymm0, 1
vpextrb eax, xmm1, 4
vpextrb ecx, xmm1, 8
vpextrb edi, xmm1, 12
vpshufb xmm0, xmm0, xmm6
vpunpckldq      xmm0, xmm0, xmm1
vpinsrb xmm0, xmm0, eax, 5
vpinsrb xmm0, xmm0, ecx, 6
vpinsrb xmm0, xmm0, edi, 7
mov     rax, qword ptr [rdx + 32]
vmovq   qword ptr [rax + rsi], xmm0
```


</details>

Reply:
Oh wow, thank you!

I've looked closer at the assembly and there are still two places with lots of instructions on `xmm` registers instead of `ymm`.

One is register spilling with lots of `vmovqa` instructions before and after calls to vec reserving more capacity, but that code is never executed because we already reserved capacity earlier, so that's not going to affect performance.

We also load all the data with `vpinsrb` element by element because of the `load3` function that I refactored to be safe but didn't optimize. It should be relatively easy to get the compiler to use vectorized loads instead, something like `vpshufb`. I'll give it a shot later today.

**Update:** I couldn't get a working version based on vector loads and shuffles that's also faster. I either end up with fast ones that make the test I wrote fail, or everything works but performance regresses. Either way the ceiling on the end-to-end improvement here is 3% and I don't think it's worth the trouble.

Reply:
@okaneco unfortunately your latest patch produces incorrect output, probably something about channel ordering.

Test image, sourced from https://commons.wikimedia.org/wiki/File:2010-07-21_Alice-in-Wonderland_by-David-Revoy.jpg and downscaled:

<img width="1009" height="1483" alt="Alice-in-Wonderland_by-David-Revoy_2010-07-21_sm" src="https://github.com/user-attachments/assets/fb9e1028-843e-4fc6-91da-a2b70030e689" />

Encoding result with your patch:

![out](https://github.com/user-attachments/assets/386ae51b-35c6-462c-a912-a66405182a7a)

This also indicates that the testing for this code is insufficient. I'll add tests to verify it against the scalar codepath.

Reply:
I'm going to leave this PR as only a refactoring for safety + adding tests. But please feel free to open a PR for further optimizations stacked on top of this one.

Reply:
Yes, that's best left to a followup, sorry for the distractions.
The scalar test is really helpful to have so at least that was a positive to come out of this.

I've pushed the commit to remove transmute here - https://github.com/vstroebel/jpeg-encoder/pull/22
I wrote this originally, but seeing that the previous patch passed tests, I assumed there was more coverage.